torch/_inductor/fx_passes/decompose_mm.py

```diff
@@ -0,0 +1,554 @@
+            mm_i, reduce_op, group_size, group_name
+        )
+        rs_out_i = torch.ops._c10d_functional.wait_tensor(w)
+        rs_outs.append(rs_out_i)
+
+    rs_out = torch.cat(rs_outs)
+    return (rs_out,)
+
+
+def _size_hint(s: sympy.Expr) -> int:
+    hint = size_hint(s)
+    if hint is not None:
+        return hint
+    return 0
+
+
+def _is_contiguous(t) -> bool:
+    return t.is_contiguous(memory_format=torch.contiguous_format)
+
+
+def _select_optimal_chunk_count(
+    size_before_split: int,
+    num_chunks: list[int],
+    min_size_after_split: int,
+    additional_check=None,
+):
+    """Pick the largest chunk count that satisfies minimum size constraint.
+
+    Returns the greatest number from num_chunks such that
+    size_before_split // chunk_count >= min_size_after_split.
+
+    Args:
+        size_before_split: Total size to be split
+        num_chunks: List of candidate chunk counts
+        min_size_after_split: Minimum size each chunk must have
+        additional_check: Optional callable(chunk_count) -> bool for extra validation
+    """
+    sorted_num_chunks = sorted(num_chunks)
+    for n in reversed(sorted_num_chunks):
+        if additional_check is not None and not additional_check(n):
+            continue
+        if size_before_split // n >= min_size_after_split:
+            return n
+    return None
+
+
+def _find_matmul_node(input_node):
+    """Walk backwards from input_node to find the matmul node.
+
+    Traverses the graph backwards through nodes with a single input,
+    collecting intermediate nodes until a matmul operation is found.
+
+    Args:
+        input_node: Starting node to walk backwards from
+
+    Returns:
+        Tuple of (matmul_node, intermediate_nodes) where:
+        - matmul_node: The aten.mm.default node, or None if not found
+        - intermediate_nodes: List of nodes between input_node and matmul_node
+    """
+    intermediate_nodes = []
+    node = input_node
+    while len(node.all_input_nodes) == 1:
+        intermediate_nodes.append(node)
+        node = node.all_input_nodes[0]
+        if node.target == aten.mm.default:
+            return node, intermediate_nodes
+    return None, []
+
+
+def _extract_post_mm_ops(intermediate_nodes):
+    """Extract and validate operations between matmul and reduce-scatter.
+
+    Validates that only supported operations appear between the matmul and
+    reduce-scatter: pointwise ops (e.g., dtype conversions) and at most one
+    2D->3D reshape/view operation.
+
+    Args:
+        intermediate_nodes: List of nodes between matmul and reduce-scatter
+
+    Returns:
+        Tuple of (post_mm_ops, view3d_node) where:
+        - post_mm_ops: List of _PostMMOp or _ViewOp to apply after each matmul chunk
+        - view3d_node: The 2D->3D view node if present, else None
+        Returns (None, None) if validation fails
+    """
+
+    def _is_pointwise(n):
+        return n.target in (torch.ops.prims.convert_element_type.default,)
```

Comment:
you can do `torch.tag.Pointwise` in getattr(n.target, "tags", ())`.  and check single input.